File: src/models/wallpaper.ts
export interface WallpaperRecord {
  id: number;
  provider: string;
  remoteId: string;
  name: string;
  sourcePageUrl?: string;
  originalUrl?: string;
  thumbnailUrl?: string;
  thumbnailLocalPath?: string;
  localPath?: string;
  width: number;
  height: number;
  category: string;
  purity: string;
  mimeType?: string;
  fileSize?: number;
  aspectRatio?: string;
  hash?: string;
  downloadStatus: string;
  favorite: boolean;
  blacklisted: boolean;
  preset: boolean;
  fileAvailability: "remote" | "available" | "temporarily_unavailable" | "missing";
  storageKind: "remote_metadata" | "user_source" | "managed_download" | "processed" | "thumbnail";
  fileCopyCount: number;
  tags: string[];
}

export interface WallpaperPage {
  items: WallpaperRecord[];
  page: number;
  pageSize: number;
  total: number;
}

export interface DuplicateFileCopy {
  wallpaperId: number;
  path: string;
  storageKind: WallpaperRecord["storageKind"];
  availability: Exclude<WallpaperRecord["fileAvailability"], "remote">;
  fileSize?: number;
}

export interface DuplicateFileGroup {
  contentHash: string;
  copies: DuplicateFileCopy[];
}

export interface ThumbnailData {
  mimeType: string;
  bytes: number[];
}

export type CatalogSort = "latest" | "random" | "name";

export interface CatalogQuery {
  keyword?: string;
  name?: string;
  category?: "all" | "nature" | "anime" | "games" | "people" | "local";
  provider?: "all" | "wallhaven" | "wikimedia_commons" | "openverse" | "art_institute_chicago" | "thegamesdb" | "local";
  /** Includes every original stored on this device without changing its provider identity. */
  locallyAvailable?: boolean;
  fileBacked?: boolean;
  downloadStatus?: string;
  fileAvailability?: WallpaperRecord["fileAvailability"];
  storageKind?: WallpaperRecord["storageKind"];
  collectionId?: number;
  favorite?: boolean;
  minWidth?: number;
  minHeight?: number;
  maxWidth?: number;
  maxHeight?: number;
  aspectRatio?: string;
  mimeType?: string;
  tags?: string[];
  includeBlacklisted?: boolean;
  sort?: CatalogSort;
  page?: number;
  pageSize?: number;
}

export type ProviderSort = "latest" | "popular" | "random";

export interface ProviderQuery {
  keyword?: string;
  category: "all" | "nature" | "anime" | "games" | "people" | "local";
  minWidth: number;
  minHeight: number;
  aspectRatio?: string;
  page: number;
  pageSize: number;
  sort: ProviderSort;
  safety: string;
  /** Restricts online aggregation to the explicit source filter selected by the user. */
  providers?: string[];
}

File: src-tauri/src/models/wallpaper.rs
use serde::{Deserialize, Serialize};

use crate::image_processing::ProcessedImage;

/// Provider-neutral metadata persisted by the local catalog.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WallpaperRecord {
    pub id: i64,
    pub provider: String,
    pub remote_id: String,
    pub name: String,
    pub source_page_url: Option<String>,
    pub original_url: Option<String>,
    pub thumbnail_url: Option<String>,
    pub thumbnail_local_path: Option<String>,
    pub local_path: Option<String>,
    pub width: u32,
    pub height: u32,
    pub aspect_ratio: Option<String>,
    pub file_size: Option<i64>,
    pub mime_type: Option<String>,
    pub category: String,
    pub purity: String,
    pub hash: Option<String>,
    pub download_status: String,
    pub favorite: bool,
    pub blacklisted: bool,
    pub preset: bool,
    pub created_at: Option<String>,
    pub synced_at: String,
    pub downloaded_at: Option<String>,
    pub last_used_at: Option<String>,
    /// Separates remote-only metadata from available, offline, and missing files.
    pub file_availability: String,
    /// Identifies whether the preferred file is user-owned or application-managed.
    pub storage_kind: String,
    /// Counts all indexed paths for duplicate-copy management.
    pub file_copy_count: u32,
    pub tags: Vec<String>,
}

/// Data accepted by the catalog upsert boundary before SQLite assigns an id.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NewWallpaper {
    pub provider: String,
    pub remote_id: String,
    pub name: String,
    pub source_page_url: Option<String>,
    pub original_url: Option<String>,
    pub thumbnail_url: Option<String>,
    pub thumbnail_local_path: Option<String>,
    pub local_path: Option<String>,
    pub width: u32,
    pub height: u32,
    pub aspect_ratio: Option<String>,
    pub file_size: Option<i64>,
    pub mime_type: Option<String>,
    pub category: String,
    pub purity: String,
    pub hash: Option<String>,
    pub perceptual_hash: Option<String>,
    pub download_status: String,
    pub preset: bool,
    pub created_at: Option<String>,
    pub author: Option<String>,
    pub license_name: Option<String>,
    pub license_url: Option<String>,
    pub synced_at: String,
    #[serde(default)]
    pub tags: Vec<String>,
}

/// Manual apply result carries the retained identity because hash dedup may replace the input id.
#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AppliedWallpaper {
    pub processed: ProcessedImage,
    pub wallpaper: WallpaperRecord,
}

/// Paginated response keeps catalog queries bounded as the metadata set grows.
#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WallpaperPage {
    pub items: Vec<WallpaperRecord>,
    pub page: u32,
    pub page_size: u32,
    pub total: u64,
}

/// One physical file copy participating in a content-hash duplicate group.
#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DuplicateFileCopy {
    pub wallpaper_id: i64,
    pub path: String,
    pub storage_kind: String,
    pub availability: String,
    pub file_size: Option<i64>,
}

/// Groups file locations without deleting or silently merging user-owned originals.
#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DuplicateFileGroup {
    pub content_hash: String,
    pub copies: Vec<DuplicateFileCopy>,
}

/// Bounded metadata filters used by every Phase 7 catalog page.
#[derive(Clone, Debug, Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct CatalogQuery {
    pub keyword: Option<String>,
    pub name: Option<String>,
    pub category: Option<String>,
    pub provider: Option<String>,
    /// Restricts results to originals that are already available on this device.
    pub locally_available: bool,
    /// Includes indexed file records even when a source is currently offline or missing.
    pub file_backed: bool,
    pub download_status: Option<String>,
    pub file_availability: Option<String>,
    pub storage_kind: Option<String>,
    pub collection_id: Option<i64>,
    pub favorite: Option<bool>,
    pub min_width: Option<u32>,
    pub min_height: Option<u32>,
    pub max_width: Option<u32>,
    pub max_height: Option<u32>,
    pub aspect_ratio: Option<String>,
    pub mime_type: Option<String>,
    pub tags: Vec<String>,
    pub include_blacklisted: bool,
    pub sort: Option<String>,
    pub page: u32,
    pub page_size: u32,
}
